### DISCUSSION TITLE
StallGuard2 Implementation

### DISCUSSION THREAD

**Discussion by vaculiko** [joshua-8/TMC5072#1] [Ideas]
Hi, I would like to help with adding the StallGuard2 feature of the TMC5072, because I want to use it on my Arduino Mega. 
I have a TMC5072-BOB v1.2 that I can use for testing.
How can I help you with this?

**Comment by joshua-8** [comment]
Hi @vaculiko,
I'm really happy to hear that you are interested in contributing to this library! Stallguard is a very interesting feature and I would like to be able to use it with this library. 
If you make a pull request that adds a function that turns on stallguard of course that would be great and I would test and accept it. Or, if you are able to create or find an example program that turns on stallguard I could modify it to work within this library. Or, if you can read the data sheet and figure out what registers are used to turn on stallguard that would help me add this feature.

A good first step would be testing this library with your Arduino Mega (just seeing if you can make a motor turn), since I've not tested this library on a Mega.

Feel free to ask me any questions.

**Reply by vaculiko** [reply]
Hi @joshua-8, I have looked into the [TMC5072 datasheet](https://www.analog.com/media/en/technical-documentation/data-sheets/TMC5072_datasheet_rev1.26.pdf), section 12.4 Homing with StallGuard is exactly what I want to implement.
I am more fluent in Python so I will try to dig up the registers using [PyTrinamic library](https://github.com/analogdevicesinc/PyTrinamic/).

One of its [issues](https://github.com/analogdevicesinc/PyTrinamic/issues/94) mentions that StallGuard (SG) flag is located instide DRV_STATUS register.
When looking into the [register definiton for TMC5072](https://github.com/analogdevicesinc/PyTrinamic/blob/master/pytrinamic/ic/TMC5072.py), there are register addresses defined on line 132 and 138 as:
```
DRV_STATUS_M1   = 0x6F
DRV_STATUS_M2   = 0x7F
```
They match the values in table 6.5 Motor Driver Registers in the datasheet, which mentions the are read only.

[Line 499](https://github.com/analogdevicesinc/PyTrinamic/blob/master/pytrinamic/ic/TMC5072.py#L499) shows the register bitfield content.  Each field is defined as a tuple consisting of ( Address, Mask, Shift ).
```       
# DRV_STATUS_M1
        SG_RESULT_M1          = (0x6F, 0x000003FF,  0)
        FSACTIVE_M1           = (0x6F, 0x00008000, 15)
        CS_ACTUAL_M1          = (0x6F, 0x001F0000, 16)
        STALLGUARD_M1         = (0x6F, 0x01000000, 24)
        OT_M1                 = (0x6F, 0x02000000, 25)
        OTPW_M1               = (0x6F, 0x04000000, 26)
        S2GA_M1               = (0x6F, 0x08000000, 27)
        S2GB_M1               = (0x6F, 0x10000000, 28)
        OLA_M1                = (0x6F, 0x20000000, 29)
        OLB_M1                = (0x6F, 0x40000000, 30)
        STST_M1               = (0x6F, 0x80000000, 31)
```

The settings needed for SG to operate are located in CoolStep (CS) configuration, registers `0x6D` and `0x7D`.
Table 6.5.2. in datasheet shows the available settings for CS and SG. I will look more into this and try to setup an example script for SG setup and calibration.

---
I have tried to run stepper motors using your code on my Arduino Mega but with no sucess. 
After some fixes the code compiles, but motors do not move. 
What motors are you using? 
What units are used in definition of the driver class?

I will post my fixes as pull request this week.

**Reply by joshua-8** [reply]
Thanks for finding all that information!

I'm sorry to hear your motor isn't turning.
I've used a few different 4 wire stepper motors, with this as an example https://www.omc-stepperonline.com/short-shaft-nema-11-bipolar-1-8deg-12ncm-16-99oz-in-0-7a-28x28x51mm-4-wires-11hs20-0704s
I've found that I have to connect my 12 volt power for the motors before plugging the Arduino into my computer. 
Are you willing to share a photo or diagram of your circuit? 
What code are you using for the initial test? 

I don't know the exact units for "raw" movements, but maybe it's number of micro steps (the number gets big (thousands) for just one rotation)?
I added the multipliers to the definitions for the class to convert to units of your choice such as rotations or mm.

I'm looking forward to hearing what you find out.
